### PR TITLE
Expose MFSDP v2 lifecycle controls to integrations

### PR DESCRIPTION
**Pull request by shjwudp**
## Summary

Split the generic experimental MFSDP integration controls out of #6949 .

This PR lets an integration replace the standard module lifecycle hooks, drive backward finalization explicitly, and discover an experimental MFSDP v2 adapter through a wrapper chain. It contains no MCore schedule-specific wiring.

## Changes

- Add `skip_forward_backward_hooks` to `fully_shard()` while preserving state-dict hooks.
- Add `pre_backward(register_final_callback=False)` for explicitly managed backward schedules.
- Extend `find_megatron_fsdp()` to recognize an adapter wrapping an experimental `FsdpModule`.
- Add direct tests for hook opt-out, explicit backward setup, and wrapper discovery.

## Split relationship

This is a prerequisite split from #6484. The MCore adapter integration will live in #6692.

## Validation

- `ruff check` on all changed Python files
- `isort` on all changed Python files
- Python bytecode compilation on all changed Python files
- `git diff --check`
- Distributed GPU unit test: pending

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py [hunk 1]
```diff
@@ -107,6 +107,7 @@ def fully_shard(
     mesh: DeviceMesh,
     placements: Placements,
     mixed_precision_policy: MixedPrecisionPolicy | None = None,
+    skip_forward_backward_hooks: bool = False,
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: Same comment as https://github.com/NVIDIA/Megatron-LM/pull/6484#discussion_r3872837216

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: Pasting Jack's rely: 

Currently this PR use fine-grained hooks replace M-FSDP v2 default hooks. Keep M-FSDP v2 default hooks encouter technical challenges:

1. 1F1B breaks down the model into sub-layer schedules, which can lead to the loss handle of some modules. For example, 1F1B maintains Attn and MLP, but loses the transformerlayer handler. This forces it to find the nearest parent FsdpModule through the attn layer to complete FSDP unshard and reshard operations. Further complicating matters, we cannot assume that fully_shard always wraps a TransformerLayer. Fine-grained hooks provide links from any FsdpModule submodule to the FsdpModule, simplifying 1F1B applications.
2. The state management (FsdpModule.phase) of 1F1B hooks differs from the default management of M-FSDP v2 because multiple micro-batch forward and backward transitions occur simultaneously. If supported, defensive programming might need to be added to the M-FSDP v2 hooks. Introduce complicate.

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: > 1F1B breaks down the model into sub-layer schedules

I've yet to understand why we can't `fully_shard` sublayers. 

> The state management (FsdpModule.phase) of 1F1B hooks differs from the default management of M-FSDP v2

This reads quite philosophical. What's the problem in practice?

Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: image

>why we can't fully_shard sublayers

One problem is **1F1B** might not even be a sublayer but rather a function dispatch/combine. This division cannot be accurately captured by `fully_shard` sublayers. In the diagram, combine, dispatch, and mlp are all part of the same module, MoELayer.

In addition, replacing fully_shard tflayer with fully_shard sublayers will increase NCCL calls and CPU overhead.

> What's the problem in practice? (state management)

I want to say that fragmented scheduling like 1F1B makes the state more complex. Wrap MoELayer with fine-grined fully_shard, this MoELayer will simultaneously in both forward and backward states.

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: > a function dispatch/combine

What's wrong with that? If they are simply functions, they shouldn't own any parameters and shouldn't/needn't be `fully_shard`ed. So the 1f1b schedule simply calls them in the original form.

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: > In addition, replacing fully_shard tflayer with fully_shard sublayers will increase NCCL calls and CPU overhead.

How so? I don't understand the NCCL-call part. I **kind of** get the CPU-overhead part but that's against "optimize for simplicity by default" in https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/distributed/fsdp/src/docs/mfsdp_design.md#development-process

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: > makes the state more complex. Wrap MoELayer with fine-grined fully_shard, this MoELayer will simultaneously in both forward and backward states.

How so? The FsdpModule.Phase is tied to each FsdpModule. Submodules don't have to agree with their parents.

**Review by wujingyue — COMMENTED** [review]
Inlined

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py (hunk 1 above)
Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: > What's wrong with that? If they are simply functions, they shouldn't own any parameters and shouldn't/needn't be fully_sharded. So the 1f1b schedule simply calls them in the original form.

Although `dispatch/combine` is a function, it uses shared experts from MoELayer for computation: [dispatch](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/moe/token_dispatcher.py#L717); [combine](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/moe/token_dispatcher.py#L872); MoELayer also uses shared experts for computation: [MoELayer.shared_experts_compute](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/moe/moe_layer.py#L525)

In principle, shared experts are members of MoELayer; `dispatch/combine` simply borrows them.

Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: > How so? I don't understand the NCCL-call part. I kind of get the CPU-overhead part but that's against "optimize for simplicity by default" in https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/distributed/fsdp/src/docs/mfsdp_design.md#development-process

More granular fully sharding means more unsharding, resharding, and `reduce_gradient_groups` calls. The TransformerLayer has four modules with parameters: `Input Layernorm`, `SelfAttention`, `pre_mlp_layernorm`, and `MLP`.

If we fully shard each of them individually, it will increase FSDP communication calls by 2 to 3 times.

> How so? The FsdpModule.Phase is tied to each FsdpModule. Submodules don't have to agree with their parents.

This will be affected by the first problem, which is: we may not be able to separate their logic, as they share parameters.

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: > In principle, shared experts are members of MoELayer; dispatch/combine simply borrows them.

Sure. Still, what's wrong with that? What I said earlier still holds:

> If dispatch/combine are simply functions, they shouldn't own any parameters and shouldn't/needn't be fully_sharded. 

So shared_experts will be unsharded by MoeLayer's pre_forward/pre_backward and resharded by MoeLayer's post_forward/post_backward. dispatch/combine, which live strictly inside the forward/backward scope of MoeLayer, will still see the valid, full weights. 

> More granular fully sharding means more unsharding, resharding, and reduce_gradient_groups calls

Sure. However, that's really the nature of **fine granularity**, isn't it? More fine-grained fully_shard means more FsdpModules and therefore more communication and therefore more NCCL calls. 

As a **potential** optimization, we could batch the allgathers prefetched at the cost of delaying the compute.

**Review by wujingyue — COMMENTED** [review]
inlined

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py (hunk 1 above)
Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: > So shared_experts will be unsharded by MoeLayer's pre_forward/pre_backward and resharded by MoeLayer's post_forward/post_backward. dispatch/combine, which live strictly inside the forward/backward scope of MoeLayer, will still see the valid, full weights.

It seems possible. If we use a fixed fully_shard strategy in 1F1B, I need to discuss it with Pingtian.

> Sure. However, that's really the nature of fine granularity, isn't it? More fine-grained fully_shard means more FsdpModules and therefore more communication and therefore more NCCL calls.

I was thinking that this part could perhaps be optimized if a fixed strategy (fully_shard TransformerLayer) were used.

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: > I was thinking that this part could perhaps be optimized if a fixed strategy (fully_shard TransformerLayer) were used.

Yep. As long as the schedule is fixed, we can "bucketize" communication more aggressively.

Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: Hey @wujingyue , I gave this a try (direct call FsdpModule hooks, not fine-grained hooks), and it works well. Here is a draft implementation: https://github.com/NVIDIA/Megatron-LM/pull/6949.

However, as mentioned previously, under 1F1B scheduling, a Transformer layer may be in both the forward and backward phases concurrently. This makes the standard M-FSDP forward/backward hooks difficult to use correctly in this scenario.

Therefore, the current approach allows users to disable the forward and backward hooks (except `post_backward`) and invoke them manually instead, as shown here:
https://github.com/NVIDIA/Megatron-LM/pull/6949/changes#diff-840c0fcc9170b02caaf50b2cb94aa96a6bb376ed27453a038c5c96dc3c7862c0R293

I have not prioritized fine-grained fully_shard-based solution because it could introduce additional edge cases and potentially cause a performance regression for now.

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: > However, as mentioned previously, under 1F1B scheduling, a Transformer layer may be in both the forward and backward phases concurrently. This makes the standard M-FSDP forward/backward hooks difficult to use correctly in this scenario.

This is a bit too vague for me to understand. Do you have an example showing the problem?

**Review by wujingyue — COMMENTED** [review]
Thanks for the update!

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py (hunk 1 above)
Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: For example, under 1F1B scheduling, a Transformer Layer can be used simultaneously as both a forward layer (`f_layer`) and a backward layer (`b_layer`) to process different data. In this case, the layer is performing both forward and backward computations concurrently, as shown in the code:

https://github.com/NVIDIA/Megatron-LM/pull/6949/changes#diff-840c0fcc9170b02caaf50b2cb94aa96a6bb376ed27453a038c5c96dc3c7862c0R260

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: We added phase tracking for supporting activation recomputation. Does this mean we can't use activation recomputation with 1f1b? I'm not saying it's a big deal, but I'm wondering your expectation on this.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py [hunk 2]
```diff
@@ -293,7 +314,22 @@ def grad_hook(_: nn.Parameter) -> None:
             if not group.requires_grad:
                 continue
             for fsdp_parameter in group.fsdp_parameters:
-                fsdp_parameter.unsharded.register_post_accumulate_grad_hook(grad_hook)
+                parameter = fsdp_parameter.unsharded
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py:317]: Does this change come from https://github.com/NVIDIA/Megatron-LM/pull/6697? Do we want to merge that first so this PR gets simpler?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py [hunk 3]
```diff
@@ -256,24 +276,25 @@ def is_root(self) -> bool:
...
-        )
-        module.register_full_backward_pre_hook(
-            lambda hooked_module, _grad_output: cast(FsdpModule, hooked_module).pre_backward()
-        )
-        if self._num_trainable_parameters == 0:
-            module.register_full_backward_hook(
-                lambda hooked_module, _grad_input, _grad_output: cast(
-                    FsdpModule, hooked_module
-                ).post_backward()
+        if not self.context.custom_forward_backward_hooks:
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py:279]: Not sure if it's needed. Can you show me the call sequence? I remember we don't even `TransformerLayer.__call__` for 1f1b.

**Review by wujingyue — COMMENTED** [review]
Thanks for the PR! I reviewed the megatron_fsdp/experimental changes. After that, I'll review the adapter and the tests.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py (hunk 1 above)
Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: Yes, we can't. The current activation recompute is based on a standard `forward n layers -> backward n layers` scheduling design. For a specialized schedule such as 1F1B, we cannot safely assume what operation will execute next.

In that case, I think it would be better to disable the current activation-recomputation and prefetch optimizations, and introduce a separate scheduler that can derive these optimizations from the actual execution order. Put differently, we should optimize only after we know the schedule, rather than relying on assumptions that are valid only for the standard forward/backward flow.

With 1F1B, a module-level “forward state” or “backward state” is not particularly meaningful. The useful scheduling units are lower-level atomic operations instead:

- `unshard`
- `reshard`

One possible approach is **trace and replay**:

1. During an initial iteration, record the sequence of `unshard` and `reshard` operations.
2. Finalize and optimize that sequence.
3. Replay the optimized schedule in later iterations.

For example, if a `reshard` is immediately followed by an `unshard` for the same module or parameter group, the two operations may be removable. The scheduler could also expose a `suggest_prefetch()` interface, allowing it to provide the correct prefetch target/order based on the traced execution schedule.

Conceptually, it could look like this:
```python
class FsdpModule:
    def pre_forward(self) -> None:
        self._unshard_parameter_groups()
        self.context.execution_runner.record_unshard(self)

        # Before the trace is finalized, use a conservative static schedule.
        # After finalization, use the trace-optimized schedule.
        prefetch_module = self.context.execution_runner.suggest_prefetch(self)
        if prefetch_module is not None:
            prefetch_module._unshard_parameter_groups()

    def post_forward(self) -> None:
        if self.context.execution_runner.should_reshard(self):
            self._reshard_parameter_groups()
            self.context.execution_runner.record_reshard(self)

    def pre_backward(self) -> None:
        # Same idea as pre_forward().
        self.context.execution_runner.record_unshard(self)
        self._unshard_parameter_groups()

        prefetch_module = self.context.execution_runner.suggest_prefetch(self)
        if prefetch_module is not None:
            prefetch_module._unshard_parameter_groups()

    def post_backward(self) -> None:
        self._reduce_gradient_groups()
        self.context.execution_runner.record_reduce_gradients(self)


for inputs, labels in dataloader:
    module.context.execution_runner.trace()

    outputs = module(inputs)
    loss = loss_fn(outputs, labels)
    loss.backward()

    module.context.execution_runner.finalize()
```

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py (hunk 3 above)
Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py:279]: This works for `MoELayer`. Call sequence: https://github.com/shjwudp/Megatron-LM/blob/b610fa8942b104063880ba2d2e2bf35406996d76/megatron/core/models/common/model_chunk_schedule_plan.py#L260

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py:279]: This looks nice. Apparently, the schedule doesn't call `TransformerLayer.__call__` at all. So why do we **need** to skip the hooks?
